File: preview/logical/python/mlir/dialects/qlx.py
# ============================================================================ #
# Copyright (c) 2026 NVIDIA Corporation & Affiliates.                          #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #
"""cudaq.mlir.dialects.qlx -- CUDA-Q Logical IR dialect module.

Re-exports the TableGen-generated op and enum bindings, plus the
native extension's attribute classes.  This is the single
user-facing module for the QLX dialect; mirrors MLIR upstream's
`mlir.dialects.arith` layout where the dialect module lives directly
under `<package>.dialects.<name>` with no intermediate namespace.

Usage::

    from cudaq.mlir.dialects import qlx as qlx_dialect

    attr = qlx_dialect.PauliAttr.get("Z", ctx)
"""

from __future__ import annotations

# Generated op + enum bindings (declare_mlir_dialect_python_bindings).
from ._qlx_ops_gen import *  # noqa: F401,F403
from ._qlx_enum_gen import *  # noqa: F401,F403

# Native extension: attribute subclasses live in the `qlx` submodule
# of the combined _qlx_ext extension.
from .._mlir_libs._qlx_ext.qlx import (  # noqa: F401
    PauliAttr, set_inherent_attr,
)

File: preview/logical/ir/lib/CAPI/Dialect/QLX.cpp
/*******************************************************************************
 * Copyright (c) 2026 NVIDIA Corporation & Affiliates.                         *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 *******************************************************************************/

#include "qlx-c/Dialect/QLX.h"

#include "mlir-c/IR.h"
#include "mlir-c/Support.h"
#include "mlir/CAPI/IR.h"
#include "mlir/CAPI/Registration.h"
#include "mlir/CAPI/Support.h"

#include "qlx/Dialect/QLX/IR/QLXAttrs.h"
#include "qlx/Dialect/QLX/IR/QLXDialect.h"
#include "qlx/Dialect/QLX/IR/QLXTypes.h"
#include "qlx/Dialect/QLX/Transforms/QLXToPBC.h"

MLIR_DEFINE_CAPI_DIALECT_REGISTRATION(QLX, qlx, qlx::QLXDialect)

bool qlxLowerToPBC(MlirModule module) {
  return mlir::succeeded(qlx::lowerToPBC(unwrap(module)));
}

//===----------------------------------------------------------------------===//
// Enum attributes
//===----------------------------------------------------------------------===//

bool qlxAttributeIsAPauli(MlirAttribute attr) {
  return llvm::isa<qlx::PauliAttr>(unwrap(attr));
}

MlirAttribute qlxPauliAttrGet(MlirContext ctx, MlirStringRef value) {
  auto sym = qlx::symbolizePauli(unwrap(value));
  if (!sym)
    return MlirAttribute{nullptr};
  return wrap(qlx::PauliAttr::get(unwrap(ctx), *sym));
}

MlirStringRef qlxPauliAttrGetValue(MlirAttribute attr) {
  auto a = llvm::cast<qlx::PauliAttr>(unwrap(attr));
  return wrap(qlx::stringifyPauli(a.getValue()));
}

MlirTypeID qlxPauliAttrGetTypeID(void) {
  return wrap(qlx::PauliAttr::getTypeID());
}
